### PR TITLE
Major refactoring of OnestopId:

### LINKED ISSUE #287: refactor Onestop ID class
will be useful for RouteStopPattern Onestop IDs (on #279) and Station/StationPlatform/StationEntrance Onestop IDs (on #256)

### PR DESCRIPTION
**Pull request by doublestranded**
top-level module that defines OnestopIdBase and subclasses, and separate initialize method
identification of onestop-ids by prefix, number of components, and soon suffixes

Resolves #287

### CODE REVIEW

**Comment by doublestranded** [comment]
Resolves #287

File: app/services/onestop_id.rb [hunk 1]
```diff
@@ -1,123 +1,123 @@
...
-    when :geohash
-      !(value =~ GEOHASH_FILTER)
-    when :name
-      !(value =~ NAME_FILTER)
-    else
-      false
+  LOOKUP = Hash[OnestopId::OnestopIdBase.descendants.map { |c| [[c::PREFIX, c::NUM_COMPONENTS], c] }]
+  LOOKUP_MODEL = Hash[OnestopId::OnestopIdBase.descendants.map { |c| [c::MODEL, c] }]
+
+  def self.lookup(string: nil, prefix: nil, num_components: 3)
```
Comment (drewda) [inline app/services/onestop_id.rb]: `handler_by_string`

File: app/services/onestop_id.rb [hunk 2]
```diff
@@ -1,123 +1,123 @@
...
+
+  def self.find(onestop_id)
+    lookup(string: onestop_id)::MODEL.find_by(onestop_id: onestop_id)
+  end
+
+  def self.find!(onestop_id)
+    lookup(string: onestop_id)::MODEL.find_by!(onestop_id: onestop_id)
+  end
+
+  def self.factory(model)
```
Comment (drewda) [inline app/services/onestop_id.rb]: `handler_by_model`

File: app/services/onestop_id.rb [hunk 3]
```diff
@@ -1,123 +1,123 @@
...
+    attr_accessor :geohash, :name
 
-  def self.create_identifier(feed_onestop_id, entity_prefix, entity_id)
-    IDENTIFIER_TEMPLATE.expand(
-      feed_onestop_id: feed_onestop_id,
-      entity_prefix: entity_prefix,
-      entity_id: entity_id
-    ).to_s
-  end
+    def initialize(string: nil, geohash: nil, name: nil)
```
Comment (drewda) [inline app/services/onestop_id.rb]: throw some `ArgumentErrors`

**Pull request merged**